### PR TITLE
chore: update the Notifications BEP for external emitters

### PR DESCRIPTION
**Pull request by mareklibra**
## Hey, I just made a Pull Request!

It updates the Notifications BEP for external emitters.

#### :heavy_check_mark: Checklist

- [ ] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: beps/0001-notifications-system/README.md [hunk 1]
```diff
@@ -178,6 +178,25 @@ Let's walk through the process of sending of a notification to a user:
...
 
+### Integration with external systems
+
+The `notifications-backend` is not meant to expose the API outside of the Backstage, only other Backstage backend services are expected to connect to through the NotificationsService.
+
+To allow external system to send messages to either individual users or broadcast, a plugin `notifications-external-backend` will be created.
+
+The purpose of the plugin is to expose new REST API for creating messages, so far no reading functionality is expected. Internally, the new API will perform
+
+- external service authentication (aligned with capabilities of the Auth service, so far just shared-secret)
```
Comment (mareklibra) [inline beps/0001-notifications-system/README.md:189]: @drodil @freben What are your suggestions here, please? The old auth service supported shared secrets only for external callers. Has this been changed recently, please?

If not, I would keep a list of shared secrets, one per external system.

File: beps/0001-notifications-system/README.md [hunk 2]
```diff
@@ -178,6 +178,25 @@ Let's walk through the process of sending of a notification to a user:
 
    7b. Rather than rendering the notification directly, the notifications frontend plugin leaves that to an extension installed by the frontend part of the backend plugin that sent the notification. This allows for a more customized notification experience which can also contain custom actions and data display. This step is not in scope for this proposal.
 
+### Integration with external systems
+
+The `notifications-backend` is not meant to expose the API outside of the Backstage, only other Backstage backend services are expected to connect through the NotificationsService.
```
Comment (Rugvip) [inline beps/0001-notifications-system/README.md:183]: Hmm, it is, it has a REST API. It just happens that other Backstage plugins should use the service instead, as a more convenient TypeScript interface. It's implemented here: https://github.com/backstage/backstage/blob/0b120df27bffb774acdd09464ce47ce071f1e74e/plugins/notifications-backend/src/service/router.ts#L278

**Comment by mareklibra** [comment]
@Rugvip , If the other-plugin-developers decide to use it
- the `origin` will be `notifications-external`, not something what is usually expected from an internal processing,
- this plugin is optional but the NotificationsService is a core-service,
- the API of the NotificationsService is more handy to be used by other services compared to composing HTTP requests on their own, IMHO

Back to our discussions at the beginning of the Notifications effort, the reasoning for keeping external callers shielded from the internal API were:
- smaller package to maintain
- the feature is optional
- the external-facing API is reduced (so far to just a single POST)
- we can potentially implement additional filters/validations/security/authorization
- have a backwards compatible proxy for the future

**Comment by drodil** [comment]
> @Rugvip , If the other-plugin-developers decide to use it
> 
> * the `origin` will be `notifications-external`, not something what is usually expected from an internal processing,

The origin should come from the token here: https://github.com/backstage/backstage/blob/master/plugins/notifications-backend/src/service/router.ts#L332, not hard-coded `notifications-external`

File: beps/0001-notifications-system/README.md (hunk 1 above)
Comment (drodil) [reply beps/0001-notifications-system/README.md:189]: See this for external tokens: https://github.com/backstage/backstage/blob/master/packages/backend-plugin-api/src/services/definitions/AuthService.ts#L39

**Review by Rugvip — CHANGES_REQUESTED** [review]
To be about more explicit with the feedback on this update: I don't think this is the way we want to go. I'm very much open to a more intentional internal/external API split, but I don't think a separate plugin is the way to go.

Any reasoning for the existence of that really applies straight to the notifications backend too. For example if it's supposed to be a more ergonomic API, then we should just make the notifications backend have that instead. Same with stability, we need to keep the API stable in the notifications backend too, at least the one intended for external consumption.

**Comment by mareklibra** [comment]
The motivation behind that was the discussion a couple of months ago where we pointed to that approach.

However, considering
- there will be an (handy) option to get the token by an external caller
- we are ok to open the API externally
there is no need for this, so closing.

**Comment by mareklibra** [comment]
@Rugvip @freben For future reference, can you please point to a procedure how to get the token by an external caller?
Or reference to an ongoing work (BEP, PR, whatever).

**Comment by Rugvip** [comment]
@mareklibra yep it's here: https://backstage.io/docs/auth/service-to-service-auth#usage-in-external-callers

That's what we're looking to improve as continued work on top of #22605 though, allowing for both more simple and hopefully more polished flows

**Comment by mareklibra** [comment]
Thanks, that will be helpful. I remember a lot of struggling when passing through that based on the docs ^^^.

**Pull request closed without merging**